In [1]:
#importing library
import pandas as pd

In [2]:
#loading data
url="https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv"
df=pd.read_csv(url,error_bad_lines=False)
df

record_id  utility_id_ferc1  report_year  \
0        f1_fuel_1994_12_1_0_7                 1         1994   
1       f1_fuel_1994_12_1_0_10                 1         1994   
2        f1_fuel_1994_12_2_0_1                 2         1994   
3        f1_fuel_1994_12_2_0_7                 2         1994   
4       f1_fuel_1994_12_2_0_10                 2         1994   
...                        ...               ...          ...   
29518  f1_fuel_2018_12_12_0_13                12         2018   
29519   f1_fuel_2018_12_12_1_1                12         2018   
29520  f1_fuel_2018_12_12_1_10                12         2018   
29521  f1_fuel_2018_12_12_1_13                12         2018   
29522  f1_fuel_2018_12_12_1_14                12         2018   

           plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0                  rockport                coal       ton        5377489.0   
1      rockport total plant                coal       ton       10486945.0   
2                    gorgas                coal       ton        2978683.0   
3                     barry                coal       ton        3739484.0   
4                 chickasaw                 gas       mcf          40533.0   
...                     ...                 ...       ...              ...   
29518    neil simpson ct #1                 gas       mcf          18799.0   
29519  cheyenne prairie 58%                 gas       mcf         806730.0   
29520     lange ct facility                 gas       mcf         104554.0   
29521       wygen 3 bhp 52%                coal       ton         315945.0   
29522       wygen 3 bhp 52%                 gas       mcf          17853.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0                   16.590                      18.59   
1                   16.592                      18.58   
2                   24.130                      39.72   
3                   23.950                      47.21   
4                    1.000                       2.77   
...                    ...                        ...   
29518                1.059                       4.78   
29519                1.050                       3.65   
29520                1.060                       4.77   
29521               16.108                       3.06   
29522                1.059                       0.00   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                             18.53                1.121  
1                             18.53                1.120  
2                             38.12                1.650  
3                             45.99                1.970  
4                              2.77                2.570  
...                             ...                  ...  
29518                          4.78                9.030  
29519                          3.65                6.950  
29520                          4.77                8.990  
29521                         14.76                1.110  
29522                          0.00               11.680  

[29523 rows x 11 columns]

In [3]:
#checking the dimension of the dataframe
df.shape

(29523, 11)

#### there are 29523 rows and 11 columns in the dataframe

In [4]:
#checking statistical summary of the data
df.describe(include="all")

record_id  utility_id_ferc1   report_year  \
count                     29523      29523.000000  29523.000000   
unique                    29523               NaN           NaN   
top     f1_fuel_2011_12_89_1_12               NaN           NaN   
freq                          1               NaN           NaN   
mean                        NaN        118.601836   2005.806050   
std                         NaN         74.178353      7.025483   
min                         NaN          1.000000   1994.000000   
25%                         NaN         55.000000   2000.000000   
50%                         NaN        122.000000   2006.000000   
75%                         NaN        176.000000   2012.000000   
max                         NaN        514.000000   2018.000000   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
count             29523               29523     29343     2.952300e+04   
unique             2315                   6         9              NaN   
top           big stone                 gas       mcf              NaN   
freq                156               11486     11354              NaN   
mean                NaN                 NaN       NaN     2.622119e+06   
std                 NaN                 NaN       NaN     9.118004e+06   
min                 NaN                 NaN       NaN     1.000000e+00   
25%                 NaN                 NaN       NaN     1.381700e+04   
50%                 NaN                 NaN       NaN     2.533220e+05   
75%                 NaN                 NaN       NaN     1.424034e+06   
max                 NaN                 NaN       NaN     5.558942e+08   

        fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
count          29523.000000               29523.000000   
unique                  NaN                        NaN   
top                     NaN                        NaN   
freq                    NaN                        NaN   
mean               8.492111                 208.649031   
std               10.600220                2854.490090   
min                0.000001                -276.080000   
25%                1.024000                   5.207000   
50%                5.762694                  26.000000   
75%               17.006000                  47.113000   
max              341.260000              139358.000000   

        fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
count                   2.952300e+04         29523.000000  
unique                           NaN                  NaN  
top                              NaN                  NaN  
freq                             NaN                  NaN  
mean                    9.175704e+02            19.304354  
std                     6.877593e+04          2091.540939  
min                    -8.749370e+02           -41.501000  
25%                     3.778500e+00             1.940000  
50%                     1.737100e+01             4.127000  
75%                     4.213700e+01             7.745000  
max                     7.964521e+06        359278.000000

#### The standard deviation and 75th percentile for the energy per unit(fuel_mmbtu_per_unit) are 10.60 and 17.01 (in two decimal places) respectively from the statistical summary above

In [5]:
#checking missing values
df.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

#### There are 180 missing values in the fuel_unit column

In [6]:
#checking the concise summary of the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


In [7]:
#using groupby to count the sum of each unique value in the fuel unit column
df.groupby('fuel_unit')['fuel_unit'].count()


fuel_unit
bbl        7998
gal          84
gramsU      464
kgU         110
mcf       11354
mmbtu       180
mwdth        95
mwhth       100
ton        8958
Name: fuel_unit, dtype: int64

#### Filling missing value: since the fuel_unit column is a categorical variable, we fill missing rows with the most occuring value (mode) which is mcf

In [8]:
#filling missing value
df[['fuel_unit']]=df[['fuel_unit']].fillna(value="mcf")

In [9]:
#checking if all missing values have been filled
df.isna().sum()

record_id                       0
utility_id_ferc1                0
report_year                     0
plant_name_ferc1                0
fuel_type_code_pudl             0
fuel_unit                       0
fuel_qty_burned                 0
fuel_mmbtu_per_unit             0
fuel_cost_per_unit_burned       0
fuel_cost_per_unit_delivered    0
fuel_cost_per_mmbtu             0
dtype: int64

#### All missing values have been filled

In [10]:
#checking for duplicate rows
df.duplicated().any()

False

#### There are no duplicates data in the dataset

In [11]:
#taking counts of all inputs for each year
df.groupby('report_year')['report_year'].count()

report_year
1994    1235
1995    1201
1996    1088
1997    1094
1998    1107
1999    1050
2000    1373
2001    1356
2002    1205
2003    1211
2004    1192
2005    1269
2006    1243
2007    1264
2008    1228
2009    1222
2010    1261
2011    1240
2012    1243
2013    1199
2014    1171
2015    1093
2016    1034
2017     993
2018     951
Name: report_year, dtype: int64

In [12]:
#grouping by the fuel type code year and printing the first entries in all the groups
df.groupby('fuel_type_code_pudl').first()

record_id  utility_id_ferc1  report_year  \
fuel_type_code_pudl                                                          
coal                  f1_fuel_1994_12_1_0_7                 1         1994   
gas                  f1_fuel_1994_12_2_0_10                 2         1994   
nuclear               f1_fuel_1994_12_2_1_1                 2         1994   
oil                   f1_fuel_1994_12_6_0_2                 6         1994   
other                f1_fuel_1994_12_11_0_6                11         1994   
waste                 f1_fuel_1994_12_9_0_3                 9         1994   

                     plant_name_ferc1 fuel_unit  fuel_qty_burned  \
fuel_type_code_pudl                                                
coal                         rockport       ton        5377489.0   
gas                         chickasaw       mcf          40533.0   
nuclear              joseph m. farley       kgU           2260.0   
oil                      clinch river       bbl           6510.0   
other                      w.f. wyman       bbl          55652.0   
waste                    b.l. england       ton           2438.0   

                     fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
fuel_type_code_pudl                                                   
coal                           16.590000                     18.590   
gas                             1.000000                      2.770   
nuclear                         0.064094                     28.770   
oil                             5.875338                     32.130   
other                           0.149719                     14.685   
waste                           0.015939                     34.180   

                     fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
fuel_type_code_pudl                                                     
coal                                       18.530                1.121  
gas                                         2.770                2.570  
nuclear                                     0.000                0.450  
oil                                        23.444                5.469  
other                                      15.090                2.335  
waste                                      34.180                1.072

#### Question 3

In [13]:
#calculating the average fuel_cost_per_unit_burned for each fuel type
df.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

#### The fuel type with the lowest average cost per unit is GAS

In [14]:
number_of_missing_values= 180
total_num_of_rows=29523
percentage_of_missing= (number_of_missing_values/total_num_of_rows)*100
print(percentage_of_missing)

0.609694136774718


#### percentage of missing rows is 0.610% (to 3 decimal places)

#### Question 5: Skewness and Kurtosis

In [15]:
#calculating skewness for fuel quantity burned
df['fuel_qty_burned'].skew()

15.851495469109503

#### the skewness for the fuel quantity burned is 15.85 (to 2dp.)

In [16]:
#calculating kurtosis for fuel quantity burned
df['fuel_qty_burned'].kurt()

651.3694501337732

#### the kurtosis for the fuel quantity burned is 651.37 (to 2dp.)

#### Question 8

In [17]:
#finding the correlation between features in the dataset
df.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

#### fuel_qty_burned and fuel_mmbtu_per_unit have the second and third lowest correlation with fuel_cost_per_unit burned

In [18]:
#calculating percentage change in fuel_cost_per_unit_burned for each year
df.groupby('report_year')['fuel_cost_per_unit_burned'].pct_change().head(5)

0         NaN
1   -0.000538
2    1.137783
3    0.188570
4   -0.941326
Name: fuel_cost_per_unit_burned, dtype: float64

##### Question 10

In [19]:
#calculating the average fuel_cost_per_unit_delivered in each year
df.groupby('report_year')['fuel_cost_per_unit_delivered'].mean()

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

#### Year 1997 has the highest average fuel cost per unit delivered